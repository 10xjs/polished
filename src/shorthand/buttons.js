// @flow
const stateMap = [undefined, null, 'active', 'focus', 'hover']

function generateButtons(state: ?string) {
  const stateSuffix = state ? `:${state}` : ''
  return `
    button${stateSuffix},
    input[type="button"]${stateSuffix},
    input[type="reset"]${stateSuffix},
    input[type="submit"]${stateSuffix}`
}

type State =
  | typeof(undefined)
  | null
  | 'active'
  | 'focus'
  | 'hover';

function buttons(...states: Array<State>) {
  if (states.length === 0) return generateButtons()
  const selectors = []
  for (let i = 0; i < states.length; i += 1) {
    if (!stateMap.includes(states[i])) throw new Error('buttons only accepts the following state: "active", "focus", or "hover". You can also pass no state at all.')
    selectors.push(generateButtons(states[i]))
  }
  return selectors.join(',')
}

export default buttons
