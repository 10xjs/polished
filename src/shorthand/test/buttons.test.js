import buttons from '../buttons'

describe('buttons', function() {
  // Single States
  it('populates all base button selectors', function() {
    expect({[buttons()]: {'border-color': 'black'}}).toMatchSnapshot()
  })
  it('populates all active button selectors', function() {
    expect({[buttons('active')]: {'border-color': 'black'}}).toMatchSnapshot()
  })
  it('populates all focus button selectors', function() {
    expect({[buttons('focus')]: {'border-color': 'black'}}).toMatchSnapshot()
  })
  it('populates all hover button selectors', function() {
    expect({[buttons('hover')]: {'border-color': 'black'}}).toMatchSnapshot()
  })
  it('throws an error when passed a state it does not recognize', function() {
    expect(
      function () { return {[buttons('clicked')]: {'border-color': 'black'}} }
    ).toThrow('buttons only accepts the following state: "active", "focus", or "hover". You can also pass no state at all.')
  })

  // Multiple States
  it('populates button selectors for two states', function() {
    expect({[buttons('active', 'hover')]: {'border-color': 'black'}}).toMatchSnapshot()
  })
  it('populates both base button selectors and button selectors with a state', function() {
    expect({[buttons(null, 'hover')]: {'border-color': 'black'}}).toMatchSnapshot()
  })
  it('throws an error when one of multiple states is not recognized', function() {
    expect(
      function () { return {[buttons('hover', 'clicked')]: {'border-color': 'black'}} }
    ).toThrow('buttons only accepts the following state: "active", "focus", or "hover". You can also pass no state at all.')
  })

})
