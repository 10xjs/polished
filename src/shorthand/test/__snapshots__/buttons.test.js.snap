exports[`buttons populates all active button selectors 1`] = `
Object {
  "button:active,
    input[type=\"button\"]:active,
    input[type=\"reset\"]:active,
    input[type=\"submit\"]:active": Object {
    "border-color": "black",
  },
}
`;

exports[`buttons populates all base button selectors 1`] = `
Object {
  "
    button,
    input[type=\"button\"],
    input[type=\"reset\"],
    input[type=\"submit\"]
    ": Object {
    "border-color": "black",
  },
}
`;

exports[`buttons populates all focus button selectors 1`] = `
Object {
  "button:focus,
    input[type=\"button\"]:focus,
    input[type=\"reset\"]:focus,
    input[type=\"submit\"]:focus": Object {
    "border-color": "black",
  },
}
`;

exports[`buttons populates all hover button selectors 1`] = `
Object {
  "button:hover,
    input[type=\"button\"]:hover,
    input[type=\"reset\"]:hover,
    input[type=\"submit\"]:hover": Object {
    "border-color": "black",
  },
}
`;

exports[`buttons populates both base button selectors and button selectors with a state 1`] = `
Object {
  "button,
    input[type=\"button\"],
    input[type=\"reset\"],
    input[type=\"submit\"]
    ,
    button:hover,
    input[type=\"button\"]:hover,
    input[type=\"reset\"]:hover,
    input[type=\"submit\"]:hover": Object {
    "border-color": "black",
  },
}
`;

exports[`buttons populates button selectors for two states 1`] = `
Object {
  "button:active,
    input[type=\"button\"]:active,
    input[type=\"reset\"]:active,
    input[type=\"submit\"]:active
    ,
    button:hover,
    input[type=\"button\"]:hover,
    input[type=\"reset\"]:hover,
    input[type=\"submit\"]:hover": Object {
    "border-color": "black",
  },
}
`;
